src/grid/CMakeLists.txt

```diff
@@ -4,8 +4,9 @@
 #!                                                                                                 !
 #!   SPDX-License-Identifier: BSD-3-Clause                                                         !
 #!-------------------------------------------------------------------------------------------------!
+cmake_minimum_required(VERSION 3.22)
 
-set(GRID_SRCS_C
```

Comment:
I repeat do not generate these files all together. CMakeLists.txt modifications should be the responsibility of the developer who want to add a functionality to it. And if PR review is enforced (which is not the case), then mistakes can be corrected during the review process. 

There is no point for generating a CMakeLists.txt with a python script as CMakeLists.txt are not that difficult to read and modify. The difficulty is in finding dependencies and compilers which cmake does most of the work for us already

- We are handwriting one script of the other. I really does not see the rational behind using a script for generating CMakeLists.txt.

- If editing a CMakeLists.txt with a text editor is tedious then editing a python script (the argument cuts both ways) that generates these files is equally tedious if not more because we have to work with string and we loose the benefit of the cmake mode provided by the editor. The amount of work is the same if not more with the python solution.

- Actually This PR already reduces the amount of code.

Reply:
Yes, complicated changes, like adding a new dependency, should be done directly in the CMake files. However, complicated changes are rare.

The much more common case is adding or renaming a file. Our old build system handled those changes completely automatically. So, I'd like the new build system be as easy to use as the old one.

Ideally we can achieve this automatization within CMake, but if that's not possible then the next best thing is a Python script.

Reply:
It is possible to generate the list with cmake (`file(GLOB ...)` syntax) I did that before reintroducing the explicit list (three lines of code for the most naive version for generating the list) but then I had to make thing complicated again because of the difference between executable/library (and it is more than three lines).

Renaming/deleting/adding a file happens once in a while and it is not complicated to search/replace a string so I really do not understand why the list should be generated automatically. 

- Compactness and readability well yes at first then things get a bit more complicated afterwards

This poses more issues than this functionality solves

- adding files automatically is dangerous when working with a dirty source tree (yes it happens often). Doing this imposes to developers to be cautious which is good but then what's the difference between this and editing a list. Cautious equally apply.
- the current tree structure imposes removing the executable source code from the list because you have to call _add_library_ for libraries and `add_executable` for executable. They are not generated the same way and do not necessarily have the same properties. There are both binary, all filles are binary it does not mean they represent the same thing and can be treated the same way.  cmake, the linker etc do not exactly behave the same way when they generate them. 
- it imposes thing like `#ifdef __FUNCTIONALITY put all the code we want #endif` because all files are compiled even when the content of the file is 100 % depending on a optional library that is turned off at compilation time (i think even fortran generates a warning or an error when the file is empty). What is in `src/grid/{gpu,hip)` with `__NO_OFFLOAD_GRID` is an example.  What about using macros in C or the imbrication of multiple interfaces before actually calling the dependency function (although I understand why it is needed sometimes). I have seen better question readability.
- we have no clue what files are actually compiled either. We might gain in compactness but we loose the information what source code is compiled and when the code is on or off. It is actually much easier to find the information "oh gpu support is on + option 1,2,3; the functionality depends on file x,y,z" in a text file than searching through the tree to find which file  is actually containing the code. Yes grep is useful but you still need to know what to search for etc.... 
- Also note that the list generation is not done in the Makefile contained in `src/grid` and `src/dbm` directories so if consistency is requested then apply the rules to everything. 
- The current build system is doing it but we do not have to always follow what was done in the past. (look at the toolchain for instance)

Reply:
> It is possible to generate the list with cmake (file(GLOB ...) syntax) ... but then I had to make thing complicated 
> again because of the difference between executable/library (and it is more than three lines).

Yes, it's indeed complicate. In the rest of your post is essentially a description of an algorithm that produces a CMakefile for a given set of source files. Currently, you are running this algorithm in your head and checking in only the result. I think, instead you should spell out this algorithm explicitly as code - using either CMake's language or Python. This would have two major advantages:
1. Anybody will be able to run the same algorithm on new files.
2. We can review and gradually improve the algorithm over time.

Reply:
I wrote what you suggest me to do before reverting to a list. It is simply not manageable to take all files and then remove the files that contains main functions for instance of should be included in libraries only. 
We still have to do `git add/mv/rm` and it is really easy to search the same string in the list and remove or rename the item.

Reply:
Also, [GLOB is discouraged](https://cmake.org/cmake/help/latest/command/file.html#filesystem) by CMake as it may break generators and adds cost to each run. Comparing the number of times people will call CMake to simply build, to the number of times new files are added, the case is quiet clear.

Reply:
Whatever way the problem is "solved" the information needs to be encoded. I reverted the PR introducing what you want to reintroduce because what was done made no sense in practice. The 1k files / 30 dependencies argument falls short as well because this information needs to be encoded somewhere. I prefer the explicit list because it is clear what's going on. Then including or excluding dependencies can be handled by adding files to the list (not removing them actually) with if statements and compilation flags which we do already.   

Only the developer creating the PR knows about this at instant t. This should be reflected both in the build system and git. No automatic generation of the cmakelists.txt should be allowed or modified without express consent of the developer for traceability reason. It means the CI/CD or the merge process should not modify any files (even formatting should be an argument for rejection which we have) outside fixing merge conflicts (which we do not have in practice). 

To take an example that will speak to many, the version between a submitted paper and the officially published one is very often different. However the soon to be published version is always sent to the authors for proof readings before publication (it is the validation process on the authors side). It is almost never done automatically in high impact journals (i speak about nature, science, APS papers here).
 
I was not really describing an algorithm to generate this list but exposing arguments against the idea. It was interpreted as an algorithm which was not the purpose of the comment. @dev-zero was pointing out that GLOB should be avoided whenever possible he was not suggesting anything else.

Reply:
> 



> The information is already encoded in the source files. The old build system relies on this. So whatever you write into the CMakefiles is redundant and therefore susceptible to inconsistencies.

- well the argument is based on the implicit vs explicit idea. coding this information directly in the source file is implicit and still error prone as it does not forbid compiling files that should not be compiled in the first place. The explicit argument (the CMakeLists.txt is the only source of trust thanks @dev-zero for pointing that out) remove this. 

- we put pressure on the preprocessor which will eventually generate empty files (which breaks C and C++ standards in theory, fortran compilers often expect some end statement at the end as well) for no real reasons. 

This is really not the topic of this PR, which was initially a bug fix and bacame a bit more than that (I had to fix typos etc...).

> I realize that writing the high-level algorithm is more work and I'm happy to do it myself. I'm certain it will pay off as we tweak and refine it over the coming months/years. Having a process for systematic improvements is the essence of software engineering.

you are more skilled in python than I do. if there is a way to do this with us agreeing about it I am sure you can get this working.

> I never suggested that it should happen automatically. The CI would [merely check](https://github.com/cp2k/cp2k/blob/74355ed86a302d19cf0c2ebc9855f0ecbd61f87e/tools/docker/scripts/test_python.sh#L31) that the committed CMakefile is consistent with what the generator would produce, i.e. it's essentially a very strict linter.
> 
> So, I would suggest we merge this PR and then I'll add a matching `cmakelists_linter.py`. What do you think?

I would support a consistency check by all means. But I will only consider the `CMakeLists.txt` as the only explicit source of truth not the source tree nor the content of the source files. People can commit file accidentally and this can result in catastrophic error along the way even in the implicit case. 

- The list should remain explicit if we want the system to be reliable. Modifying the list explicitly imply a commit (separated or not does not matter) etc... Which means we can tract it and validate it (the validation are the regtests etc...). 

- the difference between applications and libraries should also be considered. 

- if the list should be separated from the CMakeLists.txt and stored in a separate file for convenience with parsing then so be it (we can call it compiled_files_list.txt for instance and located in src). This file become defacto the **only** source of truth for the list of compiled files but no automatic generation of this file either (not even offer the possibility to do so) because it will encourage bad behavior. 

-The CMakeLists.txt should contain all cmake commands for targets etc....

Reply:
> I would support a consistency check by all means. 

Awesome :-)

> if the list should be separated from the CMakeLists.txt and stored in a separate file for convenience with parsing then so be it (we can call it compiled_files_list.txt for instance and located in src).

Yes, moving the source of truth to a config file would make the consistency check a lot easier!

> This file become defacto the only source of truth for the list of compiled files but no automatic generation of this file either (not even offer the possibility to do so) because it will encourage bad behavior.

Got it! Works for me.

> if there is a way to do this with us agreeing about it I am sure you can get this working.

I'm sure we'll find a way. I'm pretty flexible about the details. I mostly care about the ability to eliminate entire bug-classes at once.